frontend/packages/dev-console/src/components/topology2/Topology.tsx

```diff
@@ -0,0 +1,183 @@
+  const visRef = React.useRef<Visualization | null>(null);
+  const componentFactoryRef = React.useRef<ComponentFactory | null>(null);
+  const [model, setModel] = React.useState<Model>();
+  const [selectedIds, setSelectedIds] = React.useState<string[]>([]);
+
+  if (!componentFactoryRef.current) {
+    componentFactoryRef.current = new ComponentFactory(serviceBinding);
+  }
+
+  if (!visRef.current) {
+    visRef.current = new Visualization();
+    visRef.current.registerLayoutFactory(layoutFactory);
+    visRef.current.registerComponentFactory(componentFactoryRef.current.getFactory());
+    visRef.current.addEventListener<SelectionEventListener>(SELECTION_EVENT, (ids: string[]) => {
+      // set empty selection when selecting the graph
+      if (ids.length > 0 && ids[0] === graphModel.graph.id) {
+        setSelectedIds([]);
+      } else {
+        setSelectedIds(ids);
+      }
+    });
+    visRef.current.fromModel(graphModel);
+  }
+
+  React.useEffect(() => {
+    componentFactoryRef.current.serviceBinding = serviceBinding;
+  }, [serviceBinding]);
+
+  React.useEffect(() => {
+    const newModel = topologyModelFromDataModel(data);
+    visRef.current.fromModel(newModel);
+    setModel(newModel);
+    if (selectedIds.length && !visRef.current.getElementById(selectedIds[0])) {
+      setSelectedIds([]);
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [data]);
+
+  React.useEffect(() => {
+    let resizeTimeout = null;
+    if (selectedIds.length > 0) {
+      const selectedEntity = visRef.current.getElementById(selectedIds[0]);
+      if (selectedEntity && isNode(selectedEntity)) {
+        resizeTimeout = setTimeout(
+          action(() => {
+            visRef.current
+              .getGraph()
+              .panIntoView(selectedEntity, { offset: 20, minimumVisible: 40 });
+            resizeTimeout = null;
+          }),
+          500,
+        );
+      }
+    }
+    return () => {
+      if (resizeTimeout) {
+        resizeTimeout.cancel();
+      }
+    };
+  }, [selectedIds]);
+
+  const onSidebarClose = () => {
+    setSelectedIds([]);
+  };
+
+  const renderControlBar = () => {
+    return (
+      <TopologyControlBar
+        controlButtons={createTopologyControlButtons({
+          ...defaultControlButtonsOptions,
+          zoomInCallback: action(() => {
+            visRef.current.getGraph().scaleBy(4 / 3);
+          }),
+          zoomOutCallback: action(() => {
+            visRef.current.getGraph().scaleBy(0.75);
+          }),
+          fitToScreenCallback: action(() => {
+            visRef.current.getGraph().fit(80);
+          }),
+          resetViewCallback: action(() => {
+            visRef.current.getGraph().reset();
+            visRef.current.getGraph().layout();
+          }),
+          legend: false,
+        })}
+      />
+    );
+  };
+
+  const selectedItemDetails = () => {
+    const selectedEntity = selectedIds[0] ? visRef.current.getElementById(selectedIds[0]) : null;
```

Comment:
Selected item details only works for the 1st selected id?

Reply:
Framework supports multi select. Currently only using single selection.